aws/src/main/java/org/apache/iceberg/aws/LakeFormationAwsClientFactory.java

```diff
@@ -0,0 +1,153 @@
+ * otherwise it uses AssumingRole credentials.
+ */
+public class LakeFormationAwsClientFactory extends AssumeRoleAwsClientFactory {
+
+  public static final String LF_AUTHORIZED_CALLER = "LakeFormationAuthorizedCaller";
+
+  private String glueCatalogId;
+  private String dbName;
+  private String tableName;
+  private String glueAccountId;
+  private String region;
+
+  public LakeFormationAwsClientFactory() {
+  }
+
+  @Override
+  public void initialize(Map<String, String> catalogProperties) {
+    super.initialize(catalogProperties);
+    Preconditions.checkArgument(tags().stream().anyMatch(t -> t.key().equals(LF_AUTHORIZED_CALLER)),
+        "must set %s through %s", LF_AUTHORIZED_CALLER, AwsProperties.CLIENT_ASSUME_ROLE_TAGS_PREFIX);
+    this.glueCatalogId = catalogProperties.get(AwsProperties.GLUE_CATALOG_ID);
+    this.dbName = catalogProperties.get(AwsProperties.LAKE_FORMATION_DB_NAME);
+    this.tableName = catalogProperties.get(AwsProperties.LAKE_FORMATION_TABLE_NAME);
+    this.glueAccountId = catalogProperties.get(AwsProperties.GLUE_ACCOUNT_ID);
+    this.region = catalogProperties.get(AwsProperties.CLIENT_ASSUME_ROLE_REGION);
+  }
+
+  @Override
+  public S3Client s3() {
+    if (isRegisteredWithLakeFormation()) {
+      S3ClientBuilder builder = S3Client.builder()
+          .endpointOverride(URI.create(s3Endpoint()))
+          .credentialsProvider(new LakeFormationCredentialsProvider(getLakeFormationClient(), buildTableArn()))
+          .region(Region.of(region()))
+          .httpClientBuilder(UrlConnectionHttpClient.builder());
+      return builder.build();
+    } else {
+      return super.s3();
+    }
+  }
+
+  @Override
+  public KmsClient kms() {
+    if (isRegisteredWithLakeFormation()) {
+      KmsClientBuilder builder = KmsClient.builder()
+          .credentialsProvider(new LakeFormationCredentialsProvider(getLakeFormationClient(), buildTableArn()))
+          .region(Region.of(region()))
+          .httpClientBuilder(UrlConnectionHttpClient.builder());
+      return builder.build();
+    } else {
+      return super.kms();
+    }
+  }
+
+  private boolean isRegisteredWithLakeFormation() {
+    Preconditions.checkArgument(dbName != null && !dbName.isEmpty(), "db name can not be empty");
+    Preconditions.checkArgument(tableName != null && !tableName.isEmpty(), "table name can not be empty");
+
+    GetTableResponse response = glue().getTable(GetTableRequest.builder()
+        .catalogId(glueCatalogId)
+        .databaseName(dbName)
+        .name(tableName)
+        .build());
+    return response.table().isRegisteredWithLakeFormation();
+  }
+
+  private String buildTableArn() {
+    Preconditions.checkArgument(region != null && !region.isEmpty(), "region can not be empty");
+    Preconditions.checkArgument(glueAccountId != null && !glueAccountId.isEmpty(), "glueAccountId can not be empty");
+    Preconditions.checkArgument(dbName != null && !dbName.isEmpty(), "db name can not be empty");
```

Comment:
buildTableArn is always called when isRegisteredWithLakeFormation, I think we can avoid db and table null checks?